### PR TITLE
[CF-106] Fetch AdServices Token

### PR DESCRIPTION
**Pull request by beylmk**
AdServices token fetching and removal of old iAD code. Posting to the backend will be the next step.

* deprecates old automaticAppleSearchAdsAttributionCollection
* adds new automaticAdServicesAttributionTokenCollection
* fetches attribution token

There is a follow-up ticket to investigate linking, and we will probably switch to weak-linking then.

### CODE REVIEW

File: Sources/Attribution/AttributionFetcher.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionFetcher.swift:22]: Maybe
```suggestion
#if canImport(AdServices)
```

File: RevenueCat.xcodeproj/project.pbxproj
Comment (NachoSoto) [inline RevenueCat.xcodeproj/project.pbxproj:307]: Should this be weakly linked?

File: Sources/Attribution/AttributionFetcher.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionFetcher.swift:91]: I think `canImport` would help here too :)

File: Sources/Logging/Strings/StoreKitStrings.swift
Comment (NachoSoto) [inline Sources/Logging/Strings/StoreKitStrings.swift:77]: 👍🏻

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:96]: I think this will need the `tvOS`/`macCatalyst` (`macOS` too?) equivalents too

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:1846]: Maybe move this to `Deprecations.swift`?

Comment (beylmk) [reply Sources/Purchasing/Purchases.swift:1846]: is there a difference between methods in Deprecations.swift and those in the // MARK: Deprecated section of Purchases.swift?

File: Sources/Attribution/AttributionFetcher.swift
Comment (beylmk) [reply Sources/Attribution/AttributionFetcher.swift:22]: oh cool, TIL. pushing most of the linking/library stuff off until we have more info, but this one seems worth doing for now :)

File: Sources/Purchasing/Purchases.swift
Comment (beylmk) [reply Sources/Purchasing/Purchases.swift:96]: yep good call. xcode never suggests the full checks for us, do they...

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [reply Sources/Purchasing/Purchases.swift:1846]: Oh I didn't realize we still had some there. The idea was to make the main file (which is already huge) smaller by only having the current API.

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [reply Sources/Purchasing/Purchases.swift:96]: Nope, only for the currently selected target unfortunately.

File: Sources/Purchasing/Purchases.swift
Comment (beylmk) [reply Sources/Purchasing/Purchases.swift:1846]: makes sense to me! i can move the rest while i'm at it...

File: Sources/Purchasing/Purchases.swift
Comment (aboedo) [inline Sources/Purchasing/Purchases.swift:732]: maybe we could do something like 
```suggestion
    # should match platform availability in https://developer.apple.com/documentation/ad_services
    @available(iOS 14.3, macOS 11.1, macCatalyst 14.3, *)
```

File: RevenueCat.xcodeproj/project.pbxproj
Comment (NachoSoto) [reply RevenueCat.xcodeproj/project.pbxproj:307]: In case I don't get to review this again, just wanted to point out that this is probably my most critical comment. I believe this has to be weakly linked so we don't make all users of the SDK implicitly link `AdServices`.

Comment (aboedo) [reply RevenueCat.xcodeproj/project.pbxproj:307]: +1 on this, this part is a very delicate change that we should research and test out thoroughly. 
It might even be a red flag for privacy-minded folks who're looking into using our SDK. Perhaps we can do the extra work of calling this through the same "reflection" we did for iAd?

Comment (beylmk) [reply RevenueCat.xcodeproj/project.pbxproj:307]: are you both okay with me doing the linking research as the follow-up ticket i created? would definitely be before merging into main. this just feels like enough work to warrant a separate PR

Comment (aboedo) [reply RevenueCat.xcodeproj/project.pbxproj:307]: absolutely! It's only important to get the linking right before merging to main, but it can totally be done as a separate thing

Comment (beylmk) [reply RevenueCat.xcodeproj/project.pbxproj:308]: okay then that's the plan!

File: RevenueCat.xcodeproj/project.pbxproj
Comment (beylmk) [inline RevenueCat.xcodeproj/project.pbxproj:2198]: are we okay with doing this? i could make a different PurchasesDeprecation for the Purchases module and only expose this one value there... or does anyone have other ideas for avoiding the build failure/warning on calling something deprecated?

Comment (NachoSoto) [reply RevenueCat.xcodeproj/project.pbxproj:2198]: I'm not sure what you mean. Which build failure?

Comment (beylmk) [reply RevenueCat.xcodeproj/project.pbxproj:2198]: the CI build was failing because of the xcode build warning on calling the deprecated `automaticAppleSearchAdsAttributionCollection` within Purchases.swift: https://app.circleci.com/pipelines/github/RevenueCat/purchases-ios/6418/workflows/1c8d28f1-5600-428c-b11d-0e81db1c8f94/jobs/24363

we made a protocol to avoid deprecation warnings from within our unit tests

Comment (beylmk) [reply RevenueCat.xcodeproj/project.pbxproj:2198]: though the fix appears to not be working anyways. i'll un-draft this PR when it's ready...

Comment (NachoSoto) [reply RevenueCat.xcodeproj/project.pbxproj:2198]: Deprecations are warnings, obsoletions are build errors. I guess we have something that ensures the build doesn't compile with warnings?
Do we need to call the deprecated method?

Comment (beylmk) [reply RevenueCat.xcodeproj/project.pbxproj:2198]: yeah, it looks like we fail on build warnings.

and yes -- it needs to be called on de-init...

**Comment by beylmk** [comment]
on hold until we get confirmation on deprecating iAD

File: Tests/UnitTests/Attribution/AttributionPosterTests.swift
Comment (beylmk) [inline Tests/UnitTests/Attribution/AttributionPosterTests.swift:284]: i'll add back related tests in the post PR

**Comment by beylmk** [comment]
need to figure out why  build-tv-watch-and-macos is failing, but otherwise this is ready for re-review

File: Sources/Attribution/AttributionFetcher.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionFetcher.swift:87]: Nice

File: Sources/Attribution/AttributionFetcher.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionFetcher.swift:78]: I think this would be better as a property:
```suggestion
    var adServicesToken: String? {
```

File: Sources/Attribution/AttributionFetcher.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionFetcher.swift:83]: Maybe do a do/catch instead of `try?` to add the underlying error to this?
Also I wonder if it should be `Logger.appleError`

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:105]: Nit:
```suggestion
     * - ``logHandler``
```

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:423]: Formatting here is weird, did you mean to put it in a new line?
```suggestion
        if #available(iOS 14.3, macOS 11.1, macCatalyst 14.3, *) {
                    postAdServicesTokenIfNeeded()
```

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:262]: :+1

File: Tests/UnitTests/Mocks/MockAttributionFetcher.swift
Comment (NachoSoto) [inline Tests/UnitTests/Mocks/MockAttributionFetcher.swift:10]: Nit: maybe move this right on top of `adServicesToken`? I think that's what other mocks do

File: Tests/UnitTests/Networking/Backend/__Snapshots__/BackendPostAttributionDataTests/iOS12-testPostAttributesPutsDataInDataKey.1.json
Comment (NachoSoto) [inline Tests/UnitTests/Networking/Backend/__Snapshots__/BackendPostAttributionDataTests/iOS12-testPostAttributesPutsDataInDataKey.1.json:2]: Thanks for removing these too!

**Review by NachoSoto — APPROVED** [review]
Looks good! Just a few minor comments.
Let me know if I can help figuring out the watchOS issue, might be because `AdServices` isn't available, so maybe weak linking might fix that?

File: Sources/Purchasing/Purchases.swift
Comment (beylmk) [reply Sources/Purchasing/Purchases.swift:423]: oh woops yes. was on a tiny screen this day and didn't realize, good catch

**Comment by beylmk** [comment]
> Does this have to be optional?

that was me trying to weak link the framework. i'm not seeing very clear guidance on how to do that in an initial google search...

**Comment by beylmk** [comment]
> Looks good! Just a few minor comments. Let me know if I can help figuring out the watchOS issue, might be because `AdServices` isn't available, so maybe weak linking might fix that?

if you have time to look that'd be great 🙏

File: Tests/APITesters/ObjCAPITester/ObjCAPITester.xcodeproj/project.pbxproj
Comment (NachoSoto) [inline Tests/APITesters/ObjCAPITester/ObjCAPITester.xcodeproj/project.pbxproj:10]: ⚠️ I think we shouldn't do this. The fact that `APITester`s needs it is a sign that users will need to add it as well, which isn't good. I'm guessing the actual problem was this: #1554.

Comment (beylmk) [reply Tests/APITesters/ObjCAPITester/ObjCAPITester.xcodeproj/project.pbxproj:10]: agreed. promise i won't forget this before merging to main.

**Comment by beylmk** [comment]
@NachoSoto thanks for your help!! i think this finally passed everything if you have time to review :)

File: Sources/Attribution/AttributionPoster.swift
Comment (NachoSoto) [inline Sources/Attribution/AttributionPoster.swift:110]: Maybe leave this as a `TODO`? it's okay if lint fails since we're not merging this in `main` yet. That way we ensure it doesn't slip through the cracks.

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:98]: This is being implicitly marked as available for _any_ version of `tvOS` and `watchOS`. You need to add this:
```swift
@available(tvOS, unavailable)
@available(watchOS, unavailable)
```

File: Sources/Purchasing/Purchases.swift
Comment (NachoSoto) [inline Sources/Purchasing/Purchases.swift:756]: Ditto here

**Pull request merged**
